fix(app): various fixes and improvements to module setup flow

vegano1 opened the pull request:
# Overview

Various fixes and improvements to the module setup flow found while testing

Closes: [EXEC-1555](https://opentrons.atlassian.net/browse/EXEC-1555) [EXEC-1559](https://opentrons.atlassian.net/browse/EXEC-1559) [EXEC-1560](https://opentrons.atlassian.net/browse/EXEC-1560) [RQA-4207](https://opentrons.atlassian.net/browse/RQA-4207) [RQA-4208](https://opentrons.atlassian.net/browse/RQA-4208) 

<img width="1028" alt="Screenshot 2025-05-23 at 5 30 51 PM" src="https://github.com/user-attachments/assets/b43808ce-e98f-4b65-93a0-1fad35928b1d" />


## Test Plan and Hands on Testing

- [x] Make sure that the success screen shows a "Setup another module" button if there are more modules to setup.
- [x] Make sure that clicking "Setup another module" takes you back to the module setup launcher screen.
- [x] Make sure the module blinks when the module setup splash screen is reached for one module.
- [x] Make sure the module blinks again after a firmware update.
- [x] Make sure the module stops blinking at the end of the module setup flow.
- [x] Make sure the module stops blinking if the module setup flow is closed while in the SelectModule screen.

## Changelog

- Add a "Setup another module" button to the success screen, which takes you back to the "Select module" screen to set up another module if applicable.
- Add `useSendIdentifyModule` to the new hooks.tsx file so we can reuse `createLiveCommands` to send the `identifyModule` command.
- Update the `SimpleWizardBody` component to allow for null children, so we can conditionally render them.
- Fix some screens that weren't using the correct SmallButton (App) vs PrimaryButton (ODD) components based on the `isOnDevice` prop.
- Blink the module when we load the module setup launcher screen, when there's a new module.
- If the install detection fails during the module setup launcher screen, blink the module red and show the `installDetect` screen.
- Add a button to `installDetect` screen, to let you retry the setup.
- Blink the module blue when the device reconnects after a firmware update
- Stop blinking the module when the module setup is complete and the modal exits.
- Block module setup exit when a module is updating.
- Make components into functions instead of constants
- Don't return an optional Null from JSX component; let the caller deal with conditional rendering
- Don't define React components as constant mid-code, just use them in the return (render) function
- Replace console.error usage with setErrorMessage, as these errors are non-recoverable for this flow + lets us get rid of returning null.


## Review requests

- Feedback would be great, especially on changing the `SimpleWizardBody` accepting null children

## Risk assessment

Low, unreleased


[EXEC-1555]: https://opentrons.atlassian.net/browse/EXEC-1555?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1559]: https://opentrons.atlassian.net/browse/EXEC-1559?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[EXEC-1560]: https://opentrons.atlassian.net/browse/EXEC-1560?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RQA-4208]: https://opentrons.atlassian.net/browse/RQA-4208?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[RQA-4207]: https://opentrons.atlassian.net/browse/RQA-4207?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
fix(app): various fixes and improvements to module setup flow
- add setup another module button + logic to success screen
- fix some screens that werent using the correct SmallButton vs PrimaryButton
- blink module when we load the module setup launcher screen
- add useSendIdentifyModule to new utils.ts file so we can reuse createLiveCommands
- blink module red if install detection fails during module setup launcher
- blink module blue when the device re-connects after a firmware update
- stop blinking module when module setup is complete and the modal exits

vegano1 committed:
update SimpleWizardBody to allow for null children, so we can conditionally render them

vegano1 committed:
dont do dumb useEffect things :)

smb2268 commented on app/src/organisms/ModuleWizardFlows/SelectLocation.tsx:
```diff
@@ -69,7 +70,11 @@ export const SelectLocation = (
-    createMaintenanceRun({})
+    if (maintenanceRunId === null) {
+      createMaintenanceRun({}).catch(error => {
+        console.error('Could not create maintenance run', error.message)
```
Instead of logging this error, let's error out the flow if this happens. You should be able to do `props.setShowErrorMessage(error.message)`

smb2268 commented on app/src/organisms/ModuleWizardFlows/SelectModule.tsx:
```diff
@@ -56,23 +58,11 @@ export const SelectModule = (props: SelectModuleProps): JSX.Element | null => {
+    if (newModules.length === 1) {
+      sendIdentifyModule(newModules[0], true)
+    }
+  }, [])
```
I think it'd be more readable to refactor this slightly so that we have a `useEffect` triggered by a change in `selectedModule` that handles issuing the `sendIdentifyModule(newModules[0], true)`. Then the `onClick` for selecting one of the buttons can just take care of stopping any other identify command and updating which module is selected.

vegano1 replied:
I ran into some issues with this, as there are some actions that we want to keep separate, like returning to a known led state after installDetect failure or not blinking when a single module is attached on component load.

smb2268 commented on app/src/organisms/ModuleWizardFlows/index.tsx:
```diff
@@ -332,12 +301,8 @@ export const ModuleWizardFlows = (
-            attachedPipette={
-              wizardFlowBaseProps.attachedPipette as PipetteInformation
-            }
+            attachedModule={wizardFlowBaseProps.attachedModule}
```
If these don't need to be retyped, I think we can remove the lines all together and make sure they're included in `wizardFlowBaseProps`!

vegano1 replied:
They don't need to be retyped, but they need to be reassigned since the value is no longer null at this point, and for some reason the control flow analysis won't know that unless we manually do so.

smb2268 commented on app/src/organisms/ModuleWizardFlows/utils.ts:
```diff
@@ -0,0 +1,40 @@
+import { useCallback } from 'react'
```
Since this file uses hooks, it needs a `.tsx` extension instead of a `.ts` extension! Let's also rename it `hooks` instead of `utils`

smb2268 reviewed (commented):
Looking great so far! I think updating children to be optional is fine but I left a few structural comments.

vegano1 committed:
- rename utils.ts -> hooks.tsx
- block moduleSetup exit when module is updating
- stop blinking module if the flow is closed while in the SelectModule component
- show the New Module Setup screen when launching module setup from the ModuleCard component
- correctly start/stop blinking when retrying after InstallDetect failure

CaseyBatten commented:
Notes from live testing:
- The module setup flow pages all appeared in order and as expected.
- The "Setup another module" button appeared and linked back to the setup splash screen.
- Modules blinked when selected, except for the third module selected sequentially in a set of three.
- Exiting the module seutp flow stopped a module from blinking
- Modules did not stop blinking when the app transitions into setting up another module
- Module did not sotp blinking when the flow was ended through the "Finish" button.

smb2268 commented on app/src/organisms/ModuleCard/index.tsx:
```diff
@@ -292,7 +292,6 @@ export const ModuleCard = (props: ModuleCardProps): JSX.Element | null => {
     >
       {showCalModal ? (
         <ModuleWizardFlows
-          attachedModule={module}
```
I don't think this fully accomplishes what Felix requested in that slack thread, but maybe I'm not following entirely. My understanding is that if we launch the flow from the banner in the module card, we want to show the new module attached screen for that specific module - in which case, we probably still want to send the specific module to the wizard flow but we'll need to update the handling within the wizard to route to the select module screen for this use case only and showing this module only.

I think this change will show the "select module" screen for all new modules, not the specific one selected. It will also make this change for flows launched via the calibrate button in the overflow menu, which we don't want.

vegano1 replied:
You're right, I missed that case while testing.
Will revert the change and address in another pr.
Nice catch!

vegano1 committed:
- Bring back attachedModule to ModuleCard, will address in separate pr
- Make sure we are correctly changing the LED state when setting up multiple modules

vegano1 committed:
use correct button for UpdateFirmware component

vegano1 committed:
increase module update timeout from 30s to 60s

mjhuff commented on app/src/organisms/ModuleWizardFlows/CloseStackerDoor.tsx:
```diff
@@ -68,6 +70,18 @@ export const CloseDoor = (props: CloseDoorProps): JSX.Element | null => {
       })
   }
 
+  const button = isOnDevice ? (
```
Generally, we should keep components beneath return statements. In functional React, `return`s act as the render method. React isn't optimized to handle this type of "assign then render" style of component writing. Effectively, this component is re-assigned a new place in memory and re-rendered every render render cycle. 

Ideally, we have a new component here that selects between the two buttons...it's probably more practical just to do a ternary select in the render method though. Basically exactly what you do on L140 of `SelectModule.tsx`

mjhuff commented on app/src/organisms/ModuleWizardFlows/SelectLocation.tsx:
```diff
@@ -69,7 +71,11 @@ export const SelectLocation = (
   const moduleName = getModuleDisplayName(attachedModule.moduleModel)
   const handleOnClick = (): void => {
-    createMaintenanceRun({})
+    if (maintenanceRunId === null) {
```
```suggestion
    if (maintenanceRunId == null) {
```

CaseyBatten commented on app/src/organisms/ModuleWizardFlows/SelectModule.tsx:
```diff
@@ -56,24 +63,15 @@ export const SelectModule = (props: SelectModuleProps): JSX.Element | null => {
-  }
+  // Handler for when there is one module
+  useEffect(() => {
+    if (isSingleModule) {
+      setSelectedModule(newModules[0])
+      sendIdentifyModule(newModules[0], true)
```
Nice 👏 
Did this also solve the LED state situation we entered where the first few modules would enter blinking states during setup but the last module being configured would not?

vegano1 replied:
Yeah, this plus the changes in the success screen

mjhuff commented on app/src/organisms/ModuleWizardFlows/SelectModule.tsx:
```diff
@@ -56,24 +63,15 @@ export const SelectModule = (props: SelectModuleProps): JSX.Element | null => {
+      setSelectedModule(newModules[0])
+      sendIdentifyModule(newModules[0], true)
+    }
+  }, [])
```
I think we want to add `isSingleModule` to the dependency array, since this is based on a network request that could update while we are on `SelectModule`?

vegano1 replied:
good call

mjhuff commented on app/src/organisms/ModuleWizardFlows/SelectModule.tsx:
```diff
@@ -107,10 +105,12 @@ export const SelectModule = (props: SelectModuleProps): JSX.Element | null => {
+      // Start blinking module, otherwise stop if multiple are available.
+      sendIdentifyModule(selectedModule, isSingleModule)
+      // Clear the selected module
+      if (!isSingleModule) setSelectedModule(null)
```
```suggestion
      if (!isSingleModule) {
      setSelectedModule(null)
      }
```
Always braces 🤓

CaseyBatten commented on app/src/organisms/ModuleWizardFlows/SelectModule.tsx:
```diff
@@ -107,10 +105,12 @@ export const SelectModule = (props: SelectModuleProps): JSX.Element | null => {
-      sendIdentifyModule(selectedModule, false)
-      setStackerNotInstalled(false)
-      setSelectedModule(null)
+      // Start blinking module, otherwise stop if multiple are available.
+      sendIdentifyModule(selectedModule, isSingleModule)
```
Question on these changes from the last build I tested, does it have a strange interaction where it can leave a stacker in a "blinking" state after a configuration has finished if they go through this try again flow?

vegano1 replied:
No, this is fixed now, the issue was with the selectModule screen.

CaseyBatten reviewed (approved):
Looking really good. Comment below on some of the LED identification changes that have happened since I last tested.

mjhuff commented on app/src/organisms/ModuleWizardFlows/Success.tsx:
```diff
@@ -24,32 +32,83 @@ export const BODY_STYLE = css`
+  setSelectedModule: (module: AttachedModule | null) => void
+}
+
+export const Success = (props: SuccessProps): JSX.Element | null => {
```
```suggestion
export function Success(props: SuccessProps): JSX.Element  {
```
Nit, I know you didn't write this line, but I think we should probably strive to keep components with the `function foo()` syntax. Some discussion on this [here](https://opentrons.slack.com/archives/C6PF2M2LC/p1742318486373119?thread_ts=1742316055.038139&cid=C6PF2M2LC) if you're interested.

What's more important (and again, not your code) is this component doesn't return `null`, nor should any component return `null`.

mjhuff commented on app/src/organisms/ModuleWizardFlows/Success.tsx:
```diff
@@ -24,32 +32,83 @@ export const BODY_STYLE = css`
-  const button = isOnDevice ? (
-    <SmallButton onClick={handleOnClick} buttonText={t('exit')} />
+
+  const finishButton = isOnDevice ? (
```
See other comment. Within the implicit render method ideally.

mjhuff commented on app/src/organisms/ModuleWizardFlows/UpdateFirmware.tsx:
```diff
@@ -70,35 +78,46 @@ export const UpdateFirmware = (
+      setIsModuleUpdating(false)
       setInProgress(false)
       setErrorMessage(t('firmware_update_failed') as string)
       if (latestRequestId != null) dispatch(dismissRequest(latestRequestId))
```
```suggestion
      if (latestRequestId != null) {
      dispatch(dismissRequest(latestRequestId))
      }
```

mjhuff commented on app/src/organisms/ModuleWizardFlows/hooks.tsx:
This is neat.

mjhuff commented on app/src/organisms/ModuleWizardFlows/index.tsx:
```diff
@@ -66,21 +66,31 @@ export const ModuleWizardFlows = (
-        handleCleanUpAndClose={handleCleanUpAndClose}
+        isModuleUpdating={wizardFlowBaseProps.isModuleUpdating}
+        handleCleanUpAndClose={() => {
+          if (selectedModule != null) sendIdentifyModule(selectedModule, false)
```
```suggestion
          if (selectedModule != null) {
          sendIdentifyModule(selectedModule, false)
          }
```
🤓

mjhuff commented on app/src/organisms/ModuleWizardFlows/useModuleSetupWizard.ts:
```diff
@@ -141,37 +142,25 @@ export function useModuleSetupWizard(
-      })
-    }
     setIsExiting(true)
+    if (attachedModule != null) sendIdentifyModule(attachedModule, false)
```
```suggestion
    if (attachedModule != null) {
    sendIdentifyModule(attachedModule, false)
    }
```

mjhuff reviewed (approved):
LGTM. All of this seems relatively straightforward, but I don't have the full behavior context as some of your squad. Just a few cleanup comments.

vegano1 committed:
- make components into functions instead of const
- dont return optional Null from JSX component, let the caller deal with conditonal rendering
- dont define react components constants mid code, then use them in the return (render) function
- replace console.error usage with setErrorMessage as these errors are non-recoverable for this flow + lets us get rid of returning null

vegano1 committed:
cleanup

vegano1 committed:
Merge branch 'edge' into EXEC-1549-setup-another-module

vegano1 committed:
include isSingleModule in useEffect

mjhuff reviewed (approved):
Looks good, thanks for the updates
